From 70b06c9e1915fb10391314fdca899f3def5255e7 Mon Sep 17 00:00:00 2001
From: Markus Boehme <markus.boehme@chainguard.dev>
Date: Fri, 12 Dec 2025 18:34:13 +0000
Subject: [PATCH] Ignore redirection error if target already exists

Do not log an error if redirection failed because the target already
exists. This can happen when multiple iptables-wrappers processes race
each other, or the filesystem is mounted read-only (and removing the
existing symlink silently failed). There is nothing to be done in these
cases and the logs are just noise.
---
 internal/iptables/alternatives.go | 2 +-
 main.go                           | 9 ++++++++-
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/internal/iptables/alternatives.go b/internal/iptables/alternatives.go
index 655ba09..12d4bad 100644
--- a/internal/iptables/alternatives.go
+++ b/internal/iptables/alternatives.go
@@ -98,7 +98,7 @@ func (s symlinkSelector) UseMode(ctx context.Context, mode Mode) error {
 		_ = os.RemoveAll(cmdPath)
 
 		if err := os.Symlink(xtablesForModePath, cmdPath); err != nil {
-			return fmt.Errorf("creating %s symlink for mode %s: %v", cmd, modeStr, err)
+			return fmt.Errorf("creating %s symlink for mode %s: %w", cmd, modeStr, err)
 		}
 	}
 
diff --git a/main.go b/main.go
index d604c87..62f7765 100644
--- a/main.go
+++ b/main.go
@@ -39,6 +39,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"io/fs"
 	"os"
 	"os/exec"
 	"path"
@@ -68,7 +69,13 @@ func main() {
 
 	selector := iptables.BuildAlternativeSelector(sbinPath)
 	if err := selector.UseMode(ctx, mode); err != nil {
-		fmt.Fprintf(os.Stderr, "Unable to redirect iptables binaries. (Are you running in an unprivileged pod?): %s\n", err)
+		// Ignore redirection error if target already exists. This may happen when
+		// multiple iptables-wrappers are racing each other or the filesystem is
+		// mounted read-only.
+		if !errors.Is(err, fs.ErrExist) {
+			fmt.Fprintf(os.Stderr, "Unable to redirect iptables binaries. (Are you running in an unprivileged pod?): %s\n", err)
+		}
+
 		// fake it, though this will probably also fail if they aren't root
 		binaryPath = iptables.XtablesPath(sbinPath, mode)
 		args = os.Args
-- 
2.51.2

